Bluetooth: Host: Add role-specific auto PHY update options

jhedberg opened the pull request:
The BT_AUTO_PHY_UPDATE option was problematic in the sense that it didn't
    really allow any kind of fine-tuning of what automation is desired. In
    particular, it didn't allow specifying which PHY was the preferred one (it
    was hardcoded to 2M) and also didn't allow specifying role-specific
    (central vs peripheral) preferences.

    To solve this, deprecate the old option, and instead introduce
    role-specific options that allow specifying 1M/2M/Coded/None as the
    preference. The new options have slightly different defaults than the new:
    for central role 2M stays as the preference, however since it's uncommon to
    require automated changes on the peripheral side the default is set to
    None there.

Thalley commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -398,12 +398,24 @@ config BT_AUTO_PHY_UPDATE
 	  but want to rely on the remote device to initiate the procedure at its
 	  discretion or want to initiate manually.
 
+config BT_CONN_AUTO_PHY_ON_PERIPHERAL
```
`BT_CONN_AUTO_PHY_ON_PERIPHERAL` is a bit weird name I think. PERIPHERAL already implies CONN, and `BT_AUTO_PHY` would follow the naming of `BT_AUTO_PHY_UPDATE`.

Maybe something like `BT_AUTO_PHY_UPDATE_ON_PERIPHERAL` or `BT_PERIPHERAL_AUTO_PHY_UPDATE`?

Thalley commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -398,12 +398,24 @@ config BT_AUTO_PHY_UPDATE
 	default y if !BT_USER_PHY_UPDATE
 	help
-	  Initiate PHY Update Procedure on connection establishment.
+	  Initiate PHY Update Procedure on connection establishment. This will attempt
+	  to update the connection to use 2M PHY, however it doesn't actually guarantee
+	  that this is what will be used in the end.
```
We should take future PHYs into consideration. 

One possible solution would be to deprecate this, and introduce 2 new Kconfig options similar to the one you have below: One for central, and one for peripheral, and the make them `choice`s where the choices are something like 

```
None
1M
2M
Coded
```

where `None` would disable the auto PHY update. This will allow users to choose the PHY preference we send to the controller, as well as being expandable with future PHYs. 

Of course, at this point, it may just be easier to refer to the `bt_conn_le_phy_update` API function, and deprecate `BT_AUTO_PHY_UPDATE`

jhedberg replied:
> Of course, at this point, it may just be easier to refer to the bt_conn_le_phy_update API function, and deprecate BT_AUTO_PHY_UPDATE

No objections from me, but I got the feeling at least @nordicjm would oppose this opinion :) (which some reasonable justifications)

jhedberg commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -135,6 +135,10 @@ Bluetooth
 * :c:struct:`bt_conn_le_cs_config` now reports both the main and sub mode as a single parameter.
 * :c:struct:`bt_conn_le_cs_main_mode` and :c:struct:`bt_conn_le_cs_sub_mode` have been replaced
   with :c:struct:`bt_conn_le_cs_mode`.
+* :kconfig:option:`CONFIG_BT_AUTO_PHY_UPDATE` now only enables automatic PHY update for central
+  role connections by default. To enable automatic PHY update also for peripheral
+  role connections, there's :kconfig:option:`CONFIG_BT_AUTO_PHY_ON_PERIPHERAL` that should be
+  enabled.
```
Another alternative is to deprecate this and replace it with role-specific options, `CONFIG_AUTO_PHY_UPDATE_PERIPHERAL` and `CONFIG_AUTO_PHY_UPDATE_CENTRAL`. The deprecated option would default to disabled but select both of the old options. Whereas the new options would have defaults of peripheral=n and central=y.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1783,8 +1796,7 @@ static void perform_auto_initiated_procedures(struct bt_conn *conn, void *unused
 
-	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) && BT_FEAT_LE_PHY_2M(bt_dev.le.features) &&
-	    !uses_symmetric_2mbit_phy(conn)) {
+	if (do_phy_update(conn) && !uses_symmetric_2mbit_phy(conn)) {
```
It's a bit of a slippery slope though with the `choice` options, since unless we also keep a virtual option for "some auto phy update" then the above if-statement will get quite complex.

Thalley replied:
> It's a bit of a slippery slope though with the `choice` options, since unless we also keep a virtual option for "some auto phy update" then the above if-statement will get quite complex.

Indeed. Might not be trivial to implement a `do_phy_update` function, but I guess we can just move the entire thing into a function (including the block that actually performs the PHY, and skip the boolean function). Not sure if it can help much w.r.t. stripping dead code, but would probably make it easier to follow (readability)

jhedberg committed:
Bluetooth: Host: Add role-specific auto PHY update options

The BT_AUTO_PHY_UPDATE option was problematic in the sense that it didn't
really allow any kind of fine-tuning of what automation is desired. In
particular, it didn't allow specifying which PHY was the preferred one (it
was hardcoded to 2M) and also didn't allow specifying role-specific
(central vs peripheral) preferences.

To solve this, deprecate the old option, and instead introduce
role-specific options that allow specifying 1M/2M/Coded/None as the
preference. The new options have slightly different defaults than the new:
for central role 2M stays as the preference, however since it's uncommon to
require automated changes on the peripheral side the default is set to
None there.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
doc: release-notes-4.3: Document Bluetooth auto PHY update changes

Document the changes related to Bluetooth auto PHY update Kconfig options.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: samples/tests: Update auto PHY usage to recent changes

Remove deprecated Kconfig option usage, and replace it with corresponding
options which yield the same behavior as before.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: Update TX buffer calculation with auto PHY changes

Make sure to consider both the new peripheral and central Kconfig options
when calculating needed TX buffers.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

Thalley reviewed (commented):
Minor comments otherwise LGTM

jhedberg committed:
Bluetooth: tests: bsim: Fix double advertising in test_connect2

There's a regression from commit 8cfad44852845cd30336d40f61dade69ab4357db
which introduced trying to enable advertising twice. Remove the other
attempt.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: tests: bsim: Restore auto PHY config

Restore the auto-phy config for tests which seem particularly sensitive
to timings related to this.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

e-rk commented:
@jhedberg  I think this PR unintentionally changed the meaning of `BT_AUTO_PHY_UPDATE`. Previously, when the `CONFIG_BT_AUTO_PHY_UPDATE=n`, the PHY would remain at default for both central and peripheral roles.
However, after this change, disabling auto PHY update will result in peripheral remaining at default PHY, but the central role will automatically switch to 2M PHY.

To keep the same behavior as previously the kconfig would have to be changed to:
```
choice
	prompt "Auto PHY update for central role"
	depends on BT_CENTRAL
	default BT_AUTO_PHY_CENTRAL_2M if BT_AUTO_PHY_UPDATE
	default BT_AUTO_PHY_CENTRAL_NONE
```
